lua_cjson.c

```diff
@@ -698,8 +700,20 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
         len = lua_array_length(l, cfg, json);
         if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object))
             json_append_array(l, cfg, current_depth, json, len);
-        else
-            json_append_object(l, cfg, current_depth, json);
+        else {
+            int as_array = 0;
+            if (lua_getmetatable(l, -1)) {
+                luaL_getmetatable(l, CJSON_AS_ARRAY_MT);
```

Comment:
Now that makes sense to me indeed, my understanding of your comment was to store the metatable in the cjson module *with* a lightuserdata key, hence the confusion, because as you are saying, that would be quite tricky (except as an upvalue).